### PR TITLE
Top N blog post

### PR DESCRIPTION
**Pull request by Alex-Monahan**
Hey folks! 

Here is the blog on top n aggregate functions! I included some fancy friendly SQL stuff in there as well, plus a benchmark vs. a window function approach. 

Open to feedback!

I'm happy to fix any linting stuff that pops up as well.

### CODE REVIEW

File: _posts/2024-10-25-topn.md [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+| l_suppkey | recent_orders |
+|----------:|------------------------------------------------------------------------|
+| 5411      | {'l_orderkey': 2543618, 'l_partkey': 105410, 'l_suppkey': 5411, ... |
+| 5411      | {'l_orderkey': 580547, 'l_partkey': 130384, 'l_suppkey': 5411, ...  |
+| 5411      | {'l_orderkey': 3908642, 'l_partkey': 132897, 'l_suppkey': 5411, ... |
+| 90        | {'l_orderkey': 4529697, 'l_partkey': 122553, 'l_suppkey': 90, ... |
+| 90        | {'l_orderkey': 4473346, 'l_partkey': 160089, 'l_suppkey': 90, ... |
+| ...       | ...           |
+
+Well that isn't so friendly looking!
```
Comment (szarnyasg) [inline _posts/2024-10-25-topn.md:242]: Why not?

File: _posts/2024-10-25-topn.md [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+| 1880      | [4852999, 2863747, 1650084] |
+| ...       | ...                         |
+
+The `max_by` function is an aggregate function, so it takes advantage of DuckDB's fast hash aggregation rather than sorting.
+Instead of sorting by `l_shipdate`, the `max_by` function scans through the dataset just once and keeps track of the `N` highest `l_shipdate` values.
+It then returns the order id that corresponds with each of the most recent shipment dates.
+The radix sort in DuckDB must scan through the dataset once per byte, so scanning only once providing a significant speedup.
+For example, if sorting by a 64-bit integer, the sort algorithm must loop through the dataset 8 times vs. 1 with this approach!
+A simple benchmark is included below.
+
```
Comment (szarnyasg) [inline _posts/2024-10-25-topn.md:190]: There seems to be a missing link here in the text. Should a benchmark come here?

**Review by szarnyasg — CHANGES_REQUESTED** [review]

File: _posts/2024-10-25-topn.md [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+This uses the built in table function `duckdb_memory()` and includes information about memory usage as well as temporary disk usage.
+The small Python script used for benchmarking is included below the results.
+The machine used for benchmarking was an M1 MacBook Pro with 16 GB RAM.
+
+| Scale factor | Metric | `QUALIFY` | `max_by` | Performance improvement |
+|-----------:|----------:|----------:|----------:|----------:|
+| 1 | Total Time | 0.58 s | 0.24 s | 2.4× |
+| 5 | Total Time | 6.15 s | 1.26 s | 4.9× |
+| 10 | Total Time | 36.8 s | 25.4 s | 1.4× |
+| 1 | Memory Usage | 0.2 GB | 1.7 GB | 8.5× |
```
Comment (szarnyasg) [inline _posts/2024-10-25-topn.md:325]: These seem swapped

**Review by szarnyasg — CHANGES_REQUESTED** [review]

File: _posts/2024-10-25-topn.md (hunk 1 above)
Comment (Alex-Monahan) [reply _posts/2024-10-25-topn.md:242]: Oh, just that the struct is all in 1 column and long. Happy to remove it you would like! Or maybe:
"Let's make this friendlier by splitting the struct back out into separate columns!"

File: _posts/2024-10-25-topn.md (hunk 2 above)
Comment (Alex-Monahan) [reply _posts/2024-10-25-topn.md:190]: Oh, the benchmark is just the results table and the included Python script. Definitely not a formal benchmark. Happy to reword - I'll take a look and see if I can rephrase.

File: _posts/2024-10-25-topn.md (hunk 1 above)
Comment (Alex-Monahan) [reply _posts/2024-10-25-topn.md]: Edited! Let me know what you think.

File: _posts/2024-10-25-topn.md (hunk 2 above)
Comment (szarnyasg) [reply _posts/2024-10-25-topn.md:190]: My only issue is that the "benchmark below" is three screens down. I think adding a link `[benchmark below](#performance-comparisons)` would resolve this.

**Pull request merged**